### PR TITLE
Gripper testing fixes

### PR DESCRIPTION
**Pull request by peterbarker**

### CODE REVIEW

File: libraries/AP_Gripper/AP_Gripper_Servo.cpp [hunk 1]
```diff
@@ -55,3 +55,16 @@ bool AP_Gripper_Servo::grabbed() const
...
 // type-specific periodic updates:
 void AP_Gripper_Servo::update_gripper() { };
+
+bool AP_Gripper_Servo::valid() const
+{
+    if (!AP_Gripper_Backend::valid()) {
+        return false;
+    }
+    int16_t current_pwm;
+    if (!RC_Channel_aux::get_radio(RC_Channel_aux::k_gripper, current_pwm)) {
```
Comment (rmackay9) [inline libraries/AP_Gripper/AP_Gripper_Servo.cpp]: Maybe we should use RC_Channel_aux::function_assigned instead?

File: libraries/AP_Mission/AP_Mission.cpp [hunk 2]
```diff
@@ -122,17 +122,20 @@ bool AP_Mission::starts_with_takeoff_cmd()
...
+                return false;
+            }
+        } else {
+            if (!read_cmd_from_storage(cmd_index, cmd)) {
+                return false;
+            }
         }
-    }
+        cmd_index++; // in case we iterate
+    } while (!is_nav_cmd(cmd));
```
Comment (rmackay9) [inline libraries/AP_Mission/AP_Mission.cpp]: Is it possible this might never complete if there are no nav commands in the mission at all?

**Comment by peterbarker** [comment]
@rmackay9 Way to spoil a good copy-and paste :-(

Fixed ;-)

File: libraries/AP_Mission/AP_Mission.cpp (hunk 2 above)
Comment (peterbarker) [inline libraries/AP_Mission/AP_Mission.cpp]: cmd_index is always incremented.

The first clause in the `if` block makes sure the mission list isn't empty.

At some stage read_cmd_from_storage will fail as cmd_index gets high enough.

I think its good.

File: ArduCopter/commands_logic.cpp [hunk 3]
```diff
@@ -802,6 +802,7 @@ bool Copter::verify_payload_place()
             g2.gripper.release();
         } else {
             gcs_send_text_fmt(MAV_SEVERITY_INFO, "Gripper not valid");
+            nav_payload_place.state = PayloadPlaceStateType_Ascending_Start;
```
Comment (OXINARF) [inline ArduCopter/commands_logic.cpp]: Unless we put a break after this, it isn't doing anything as it will be overwritten immediately after.

**Comment by OXINARF** [comment]
What's the reason behind requiring the servo gripper to have a aux switch assigned?

**Comment by peterbarker** [comment]
@OXINARF these are the output channels, not the input channels.  I hope.

File: ArduCopter/commands_logic.cpp (hunk 3 above)
Comment (peterbarker) [inline ArduCopter/commands_logic.cpp]: Fixed.

**Comment by OXINARF** [comment]
> @OXINARF these are the output channels, not the input channels. I hope.

Ah, you are correct!

Have you missed my comment about *get_next_nav_cmd*?

**Comment by peterbarker** [comment]
On Tue, 20 Dec 2016, Francisco Ferreira wrote:

> Have you missed my comment about get_next_nav_cmd?

Didn't seem to be on github when I looked, so I guessed (incorrectly) it 
has been deleted.

The change as-is is a minimum-change-to-get-it-working patch.

I've pushed up another patch which implements the function in terms of 
get_next_nav_cmd

**Comment by rmackay9** [comment]
Squashed and merged, thanks!

**Pull request closed without merging**